Check imageURL null for Element::parserSetAttributes inconsistency issue

Constellation opened the pull request:
#### d113ccd097082b7db803786051566b6abd7ceffe
<pre>
Check imageURL null for Element::parserSetAttributes inconsistency issue
<a href="https://bugs.webkit.org/show_bug.cgi?id=271873">https://bugs.webkit.org/show_bug.cgi?id=271873</a>
<a href="https://rdar.apple.com/125543680">rdar://125543680</a>

Reviewed by Ryosuke Niwa.

Element::parserSetAttributes first set all attributes into ElementRareData. And after that, it calls Element::attributeChanged.
This can cause inconsistency in each element which configures its members in attributeChanged: while attribute is already in the ElementRareData,
corresponding attributeChanged is not called *yet*. We can hit a nullptr crash in ImageLoader since it is reading src attr of &lt;image&gt; (which can be set already),
while it is also reading m_currentURL (which will be configured via attributeChanged, and it may not be called *yet*.).
In this patch, we just handle this as the same to the case we see nullptr src attr. This is fine since we call attributeChanged for each attribute so eventually
HTMLImageElement &amp; ImageLoader state converges and it becomes consistent.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):

Canonical link: <a href="https://commits.webkit.org/276826@main">https://commits.webkit.org/276826@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13185bcad2af6b511d0d8b5826a312dda1d54574

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45822 "Failed to checkout and rebase branch from PR 26600") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48405 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48492 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41859 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48128 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22348 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/46400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/22022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/39522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/18677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/40617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3866 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/40966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50256 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20814 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22118 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/43503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22477 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6384 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21808 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -218,9 +218,17 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+                // It is possible that attributes are bulk-set via Element::parserSetAttributes. In that case, it is possible that attribute vectors are already configured,
+                // but corresponding attributeChanged is not called yet. This causes inconsistency in HTMLImageElement. Eventually, we will get the consistent state, but
+                // if "src" attributeChanged is not called yet, imageURL can be null and it does not work well for ResourceRequest.
+                // In this case, we should behave as the same to attr.isNull() case.
```
should behave same as attr.isNull() or should behave as if attr.isNull()?

Constellation replied:
Done.

Constellation committed:
Check imageURL null for Element::parserSetAttributes inconsistency issue
https://bugs.webkit.org/show_bug.cgi?id=271873
rdar://125543680

Reviewed by Ryosuke Niwa.

Element::parserSetAttributes first set all attributes into ElementRareData. And after that, it calls Element::attributeChanged.
This can cause inconsistency in each element which configures its members in attributeChanged: while attribute is already in the ElementRareData,
corresponding attributeChanged is not called *yet*. We can hit a nullptr crash in ImageLoader since it is reading src attr of <image> (which can be set already),
while it is also reading m_currentURL (which will be configured via attributeChanged, and it may not be called *yet*.).
In this patch, we just handle this as the same to the case we see nullptr src attr. This is fine since we call attributeChanged for each attribute so eventually
HTMLImageElement & ImageLoader state converges and it becomes consistent.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):

Canonical link: https://commits.webkit.org/276826@main
